sprockets/mixins/http/__init__.py

```diff
@@ -6,50 +6,248 @@
 
 LOGGER = logging.getLogger(__name__)
 
 CONTENT_TYPE_JSON = headers.parse_content_type('application/json')
 CONTENT_TYPE_MSGPACK = headers.parse_content_type('application/msgpack')
+AVAILABLE_CONTENT_TYPES = [CONTENT_TYPE_JSON, CONTENT_TYPE_MSGPACK]
 DEFAULT_USER_AGENT = 'sprockets.mixins.http/{}'.format(__version__)
 
 
-HTTPResponse = collections.namedtuple(
-    'HTTPResponse',
-    ['ok', 'code', 'headers', 'body', 'raw', 'attempts', 'duration'])
-"""Response in the form of a :class:`~collections.namedtuple` returned from
-:meth:`~sprockets.mixins.http.HTTPClientMixin.http_fetch` that provides a
-slightly higher level of functionality than Tornado's
-:class:`tornado.httpclient.HTTPResponse` class.
+class HTTPResponse:
+    """Encapsulate the response(s) for requests made using the
+     :meth:`~sprockets.mixins.http.HTTPClientMixin.http_fetch` method.
 
-:param bool ok: The response status code was between 200 and 308
-:param int code: The HTTP response status code
-:param dict headers: The HTTP response headers
-:param mixed body: The deserialized HTTP response body if available/supported
-:param tornado.httpclient.HTTPResponse raw: The original Tornado HTTP
-    response object for the request
-:param int attempts: The number of HTTP request attempts made
-:param float duration: The total duration of time spent making the request(s)
+    """
 
-"""
+    def __init__(self):
+        self._exceptions = []
+        self._finish = None
+        self._json = transcoders.JSONTranscoder()
+        self._msgpack = transcoders.MsgPackTranscoder()
+        self._responses = []
+        self._start = time.time()
+
+    def __len__(self):
+        """Return the length of the exception stack and response stack.
+
+        :rtype: int
+
+        """
+        return len(self._exceptions) + len(self._responses)
+
+    def append_exception(self, error):
+        """Append an exception raised when making a request
+
+        :param Exception error: The exception raised when making the request
+
+        """
+        self._exceptions.append(error)
+
+    def append_response(self, response):
+        """Append the response to the stack of responses.
+
+        :param tornado.httpclient.HTTPResponse response: The HTTP response
+
+        """
+        self._responses.append(response)
+        if 'Warning' in response.headers:
+            LOGGER.warning(
+                'HTTP %s %s Warning (%s): %s (attempt %s)',
+                response.request.method, response.request.url,
+                response.code, response.headers['Warning'],
+                len(self._responses))
+
+    def finish(self):
+        """Mark the processing as finished"""
+        self._finish = time.time()
+
+    @property
+    def attempts(self):
+        """Return the number of HTTP attempts made by calculating the number
+        of exceptions and responses the object contains.
+
+        :rtype: int
+
+        """
+        return len(self)
+
+    @property
+    def body(self):
```

Comment:
Consider decorating this with `@functools.lru_cache(1)` to avoid decoding multiple times.